Rewrite mkldnn conv bn fuse pass tester

HulekJakub opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
Others

### PR changes
Others

### Describe
Deleted old C++ conv activation mkldnn fuse pass test and updated python one

Currently in PaddlePaddle there are 3 unit-test frameworks - one written in C++ and 2 in Python.
Old fuse pass tests (C++) fuse_pass_tester.cc are checking only if graph has correct number of / type of vertices after fusion. It doesn't check data. So these tests pass even if accuracy drops.
New python framework is superior because it checks both graph fusion and accuracy after fusion.

This change adds dedicated test for onednn conv_bn_fuse_pass 
It also fixes issue when model is using mkldnn and convolution has bias. 

Previous pass (when executed with mkldnn and the convolution had bias) merged both layers into one convolution. The problem was that while doing this it did not scale the biases properly. This implementation solves it.

HulekJakub committed:
New onednn test

HulekJakub committed:
checkopoint

HulekJakub committed:
added new test, fixed issue with onednn bias

HulekJakub committed:
fix bias check

Silv3S commented on python/paddle/fluid/tests/unittests/ir/inference/test_onednn_conv_bn_fuse_pass.py:
```diff
@@ -0,0 +1,174 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```

Silv3S commented on python/paddle/fluid/tests/unittests/ir/inference/test_onednn_conv_bn_fuse_pass.py:
```diff
@@ -0,0 +1,174 @@
+        var_shape = [out_channel]
+        mean_shape = [out_channel]
+
+        def generate_conv2d_Input():
+            return np.random.random(x_shape).astype(np.float32)
+
+        def generate_conv2d_Filter():
+            return np.random.random(w_shape).astype(np.float32)
+
+        def generate_conv2d_Bias():
+            return np.random.random(bias_shape).astype(np.float32)
+
+        def generate_bn_Scale():
+            return np.random.random(scale_shape).astype(np.float32)
+
+        def generate_bn_Bias():
+            return np.random.random(bias_shape).astype(np.float32)
+
+        def generate_bn_Mean():
+            return np.random.random(mean_shape).astype(np.float32)
+
+        def generate_bn_Var():
+            return np.random.random(var_shape).astype(np.float32)
```
maybe create one method with shape passed as argument

Silv3S commented on python/paddle/fluid/tests/unittests/ir/inference/test_onednn_conv_bn_fuse_pass.py:
```diff
@@ -0,0 +1,174 @@
+                data_gen=partial(generate_conv2d_Bias)
+            )
+
+        print(conv2d_op.inputs)
+        print(program_config.weights)
```
```suggestion
        print(conv2d_op.inputs)
        print(program_config.weights)
```

Silv3S commented on python/paddle/fluid/tests/unittests/ir/inference/test_onednn_conv_bn_fuse_pass.py:
```diff
@@ -0,0 +1,174 @@
+
+
+class TestOneDNNConvBnFusePass(PassAutoScanTest):
+    def is_program_valid(self, program_config: ProgramConfig) -> bool:
+        return True
```
```suggestion
    def is_program_valid(self, program_config: ProgramConfig) -> bool:
        return True
```
No need to overload this

Silv3S commented on paddle/fluid/framework/ir/conv_bn_fuse_pass.cc:
```diff
@@ -359,73 +359,95 @@ void ConvBNFusePass::ApplyImpl(ir::Graph* graph) const {
             conv_bias_names.size(),
             1UL,
-            platform::errors::InvalidArgument("Find input var Bais error."));
+            platform::errors::InvalidArgument("Find input var Bias error."));
```
```suggestion
            phi::errors::InvalidArgument("Find input var Bias error."));
```

Silv3S commented on paddle/fluid/framework/ir/conv_bn_fuse_pass.cc:
```diff
@@ -359,73 +359,95 @@ void ConvBNFusePass::ApplyImpl(ir::Graph* graph) const {
-        eigen_conv_bias += EigenVector<float>::From(*eltwise_y_in_tensor);
+        PADDLE_ENFORCE_EQ(conv_bias_tensor->dims(),
+                          bn_bias_tensor->dims(),
+                          platform::errors::InvalidArgument(
```
```suggestion
                          phi::errors::InvalidArgument(
```

HulekJakub committed:
remove prints, refactor code

HulekJakub committed:
delete old test

HulekJakub committed:
update python tests cmake

Silv3S reviewed (dismissed):
LGTM

HulekJakub commented:
@XieYunshen could you approve? I deleted obsolete test and fixed fuse pass.
 '::mutable_data()' was there already and I just moved it. I don't have access to DeviceContext in this class so I can not refactor it. Thank you

XieYunshen reviewed (dismissed):
LGTM

HulekJakub commented:
@YuanRisheng could you approve?
`::mutable_data()` was there already and I just moved it. I don't have access to DeviceContext in this class so I can not refactor it. Thank you

HulekJakub committed:
Merge develop and resolve conflicts

HulekJakub commented:
@Silv3S @YuanRisheng @XieYunshen I had to resolve conflicts, could you approve once again?

HulekJakub committed:
Merge branch 'PaddlePaddle:develop' into Rewrite_mkldnn_conv_bn_fuse_pass_tester

HulekJakub committed:
Delete depracated conv bias

HulekJakub committed:
Delete outdated bias from convolution test

HulekJakub commented:
@YuanRisheng @XieYunshen I had to resolve conflicts, could you approve once again?

HulekJakub commented:
@XieYunshen please approve changes in tests

jczaja reviewed (approved):
LGTM
